Sources/Core/Futures/FutureType.swift

```diff
@@ -0,0 +1,122 @@
+import Foundation
+import Dispatch
+
+/// A future result type.
+/// Concretely implemented by `Future<T>`
+public protocol FutureType {
+    associatedtype Expectation
+    func completeOrAwait(on queue: DispatchQueue, callback: @escaping ResultCallback)
+}
+
+// MARK: Convenience
+
+extension FutureType {
+    /// This future's result type.
+    public typealias Result = FutureResult<Expectation>
+
+    /// Callback for accepting a result.
+    public typealias ResultCallback = ((Result) -> ())
+
+    /// Callback for accepting the expectation.
+    public typealias ExpectationCallback = ((Expectation) -> ())
+
+    /// Callback for accepting an error.
+    public typealias ErrorCallback = ((Error) -> ())
+
+    /// Adds a handler to be asynchronously executed on
+    /// completion of this future.
+    ///
+    /// Will *not* be executed if an error occurrs
+    public func then(on queue: DispatchQueue, callback: @escaping ExpectationCallback) -> Self {
```

Comment:
`.then` not being `@discardableResult` will make the API really ugly and annoying.

Many times you don't need to catch errors, you only care about the success